### PR TITLE
[Merged by Bors] - Skip `drop` when `needs_drop` is `false`

### PR DESCRIPTION
**Pull request by DJMcNab**
# Objective

- We do a lot of function pointer calls in a hot loop (clearing entities in render). This is slow, since calling function pointers cannot be optimised out. We can avoid that in the cases where the function call is a no-op.
- Alternative to https://github.com/bevyengine/bevy/pull/2897
- On my machine, in `many_cubes`, this reduces dropping time from ~150μs to ~80μs.

## Solution
-  Make `drop` in `BlobVec` an `Option`, recording whether the given drop impl is required or not.
- Note that this does add branching in some cases - we could consider splitting this into two fields, i.e. unconditionally call the `drop` fn pointer.
- My intuition of how often types stored in `World` should have non-trivial drops makes me think that would be slower, however. 

N.B. Even once this lands, we should still test having a 'drop_multiple' variant - for types with a real `Drop` impl, the current implementation is definitely optimal.

### CODE REVIEW

File: crates/bevy_ecs/src/component.rs [hunk 1]
```diff
@@ -168,7 +169,7 @@ pub struct ComponentDescriptor {
     layout: Layout,
     // SAFETY: this function must be safe to call with pointers pointing to items of the type
     // this descriptor describes.
-    drop: for<'a> unsafe fn(OwningPtr<'a>),
+    drop: Option<for<'a> unsafe fn(OwningPtr<'a>)>,
```
Comment (bjorn3) [inline crates/bevy_ecs/src/component.rs:172]: Can you add the `// None if the underlying type doesn't need to be dropped` comment here too?

File: crates/bevy_render/src/lib.rs [hunk 2]
```diff
@@ -169,7 +169,7 @@ impl Plugin for RenderPlugin {
                         .with_system(PipelineCache::process_pipeline_queue_system)
                         .with_system(render_system.exclusive_system().at_end()),
                 )
-                .add_stage(RenderStage::Cleanup, SystemStage::parallel())
+                .add_stage(RenderStage::Cleanup, SystemStage::single_threaded())
```
Comment (bjorn3) [inline crates/bevy_render/src/lib.rs:172]: Can you revert the changes in this file?

File: crates/bevy_ecs/src/storage/blob_vec.rs [hunk 3]
```diff
@@ -375,8 +385,8 @@ mod tests {
     #[test]
     fn resize_test() {
         let item_layout = Layout::new::<usize>();
-        let drop = drop_ptr::<usize>;
-        let mut blob_vec = unsafe { BlobVec::new(item_layout, drop, 64) };
+        // usize doesn't need dropping
```
Comment (bjorn3) [inline crates/bevy_ecs/src/storage/blob_vec.rs:388]: This changes the behavior though. Maybe keep the drop function here and add a new test with `None` as drop?

**Review by bjorn3 — CHANGES_REQUESTED** [review]

File: crates/bevy_ecs/src/storage/blob_vec.rs (hunk 3 above)
Comment (DJMcNab) [reply crates/bevy_ecs/src/storage/blob_vec.rs:388]: It's not clear to me how this has different behaviour. Could you clarify this please?

Either we 'drop' the usize values by calling a function pointer, which does nothing, or we do nothing without calling a function pointer.

File: crates/bevy_render/src/lib.rs [hunk 4]
```diff
@@ -273,6 +273,11 @@ impl Plugin for RenderPlugin {
                         .get_stage_mut::<SystemStage>(&RenderStage::Cleanup)
                         .unwrap();
                     cleanup.run(&mut render_app.world);
+                }
+                {
+                    #[cfg(feature = "trace")]
+                    let _stage_span =
```
Comment (DJMcNab) [inline crates/bevy_render/src/lib.rs:279]: @bjorn3 I've left this in for now, as I'm struggling to find a compelling argument for running `Cleanup` and clearing the entities to have no differentiation. Perhaps it should be nested, but that's a different question to what you were requesting.

**Review by james7132 — APPROVED** [review]
LGTM other than the existing review comments. I was a bit suspicious of adding a branch in the hotpath for deleting/moving entities around, but it makes sense that a branch would be preferable to a what is effectively a virtual function call that most POD components don't use.

File: crates/bevy_ecs/src/storage/blob_vec.rs (hunk 3 above)
Comment (bjorn3) [reply crates/bevy_ecs/src/storage/blob_vec.rs:388]: This change means you are not testing the same as before. The drop path is no longer tested. There should be no observable difference if there are no bugs, but a test is meant to check that there are indeed no bugs.

Comment (DJMcNab) [reply crates/bevy_ecs/src/storage/blob_vec.rs:388]: This test isn't trying to test dropping though, it's testing that resizing upon `push` works. If you like, this is already testing a no-op drop, and now continues to test a no-op drop.
The case where `Drop` is `Some` is tested thoroughly below; this strictly tests more paths.

**Review by bjorn3 — APPROVED** [review]

File: crates/bevy_ecs/src/component.rs [hunk 5]
```diff
@@ -111,7 +112,11 @@ impl ComponentInfo {
     }
 
     #[inline]
-    pub fn drop(&self) -> unsafe fn(OwningPtr<'_>) {
+    /// Get the function which should be called to clean up values, i.e.
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/component.rs:115]: Wording is unclear here. Maybe "typically [`Drop`]" or something.

File: crates/bevy_ecs/src/component.rs [hunk 6]
```diff
@@ -111,7 +112,11 @@ impl ComponentInfo {
...
+    /// [`Drop`] them.
+    /// Returns `None` if the underlying type doesn't need to be dropped,
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/component.rs:117]:
```suggestion
    ///
    /// Returns `None` if the underlying type doesn't need to be dropped,
```

**Review by alice-i-cecile — APPROVED** [review]
